agent/proxycfg/snapshot.go

```diff
@@ -108,6 +124,30 @@ func (c *configSnapshotMeshGateway) IsEmpty() bool {
 		len(c.ConsulServers) == 0
 }
 
+type configSnapshotIngressGateway struct {
+	ConfigSnapshotUpstreams
+	// Upstreams is a list of upstreams this ingress gateway should serve traffic to. This is
+	// constructed from the ingress-gateway config entry, Config, and the
+	// ServiceLists fields.
+	Upstreams []structs.Upstream
+
+	// WatchedDiscoveryChains is a map of upstream.Identifier() -> CancelFunc's
+	// in order to cancel any watches when the ingress gateway configuration is
+	// changed.
```

Comment:
It might be worth mentioning why `connect-proxy` doesn't need this already. IIRC that's because we garbage collect the entire wrapper `Context` when the agent registration for a service changes, but I suppose for ingress you incrementally modify it.

Reply:
Added a comment explaining why this is necessary for ingress gateways. Felt better to structure the comment around ingress-gateways since it is a field on the ingress-specific struct